File: day_8_stock_buy_and_sell_max_one_transaction_allowed.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maximumProfit(vector<int> &prices) {
        int maxPro = 0;
        int n = prices.size();
        int minPrice = INT_MAX;
    
        for (int i = 0; i < prices.size(); i++) {
            minPrice = min(minPrice, prices[i]);
            maxPro = max(maxPro, prices[i] - minPrice);
        }
        
        return maxPro;
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> prices;
    
    for (int i = 0; i < n; i++) {
        int x;
        cin >> x;
        prices.push_back(x);
    }
    
    Solution sol;
    cout << sol.maximumProfit(prices) << endl;
    
    return 0;
}

File: Day8StockBuyAndSellMaxOneTransactionAllowed.java
import java.util.Scanner;

public class Day8StockBuyAndSellMaxOneTransactionAllowed {
    private class Solution {
        public int maximumProfit(int prices[]) {
            int maxPro = 0;
            int n = prices.length;
            int minPrice = Integer.MAX_VALUE;
            
            for(int i = 0; i < n; i++) {
                minPrice = Math.min(minPrice, prices[i]);
                maxPro = Math.max(maxPro, prices[i] - minPrice);
            }
            
            return maxPro;
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] prices = new int[n];

        for(int i = 0; i < n; i++) {
            prices[i] = scanner.nextInt();
        }

        Solution sol;
        sol = new Day8StockBuyAndSellMaxOneTransactionAllowed().new Solution();
        System.out.println(sol.maximumProfit(prices));

        scanner.close();
    }
}
